pir/cpp/server.cpp

```diff
@@ -120,24 +125,52 @@ StatusOr<std::vector<seal::Ciphertext>> PIRServer::oblivious_expansion(
     const size_t two_power_j = (1 << j);
     for (size_t k = 0; k < two_power_j; ++k) {
       auto c0 = results[k];
-      seal::Ciphertext c1;
-      // TODO: not sure which is faster: substitution operator, or multiply by
-      // factor of x? We can do one of them only once depending on order.
-      multiply_power_of_x(c0, -two_power_j, c1);
 
-      results[k] = c0;
       RETURN_IF_ERROR(substitute_power_x_inplace(
           c0, (poly_modulus_degree >> j) + 1, gal_keys));
-      context_->Evaluator()->add_inplace(results[k], c0);
 
-      results[k + two_power_j] = c1;
-      RETURN_IF_ERROR(substitute_power_x_inplace(
-          c1, (poly_modulus_degree >> j) + 1, gal_keys));
+      // This essentially produces what the paper calls c1
+      multiply_inverse_power_of_x(results[k], two_power_j,
+                                  results[k + two_power_j]);
+
+      // Do the multiply by power of x after substitution operator to avoid
+      // having to do the substitution operator a second time, since it's about
+      // 20x slower. Except that now instead of multiplying by x^(-2^j) we have
+      // to do the substitution first ourselves, producing
+      // (x^(N/2^j + 1))^(-2^j) = 1/x^(2^j * (N/2^j + 1)) = 1/x^(N + 2^j)
+      seal::Ciphertext c1;
+      multiply_inverse_power_of_x(c0, poly_modulus_degree + two_power_j, c1);
+
+      context_->Evaluator()->add_inplace(results[k], c0);
       context_->Evaluator()->add_inplace(results[k + two_power_j], c1);
     }
   }
   results.resize(num_items);
   return results;
 }
 
+StatusOr<std::vector<seal::Ciphertext>> PIRServer::oblivious_expansion(
+    const std::vector<seal::Ciphertext>& cts, size_t num_items,
+    const seal::GaloisKeys& gal_keys) const {
+  const auto poly_modulus_degree =
+      context_->Parameters()->GetEncryptionParams().poly_modulus_degree();
+
+  if (cts.size() != num_items / poly_modulus_degree + 1) {
+    return InvalidArgumentError(
+        "Number of ciphertexts doesn't match number of items for oblivious "
+        "expansion.");
+  }
+
+  std::vector<seal::Ciphertext> results;
+  results.reserve(num_items);
```

Comment:
If num_items is not a multiple of poly_modulus_degree, during the last insert, you will induce a resize and ttherefore a copy. I would suggest to reserve more, and resize at the end.

Reply:
At the end of the single-CT oblivious_expansion it resizes to the num_items for just that CT, so I think we avoid that case?